File: Data_Structure/PartialPersistentArray.cpp
// verified with:
// https://atcoder.jp/contests/code-thanks-festival-2017-open/submissions/40153584
// https://atcoder.jp/contests/agc002/submissions/40153598

#include <bits/stdc++.h>
using namespace std;

template<typename T>
struct PartialPersistentArray {
  vector<vector<T>> array;
  vector<vector<int>> times;
 
  PartialPersistentArray(){}
  explicit PartialPersistentArray(vector<T> &arr, int t = 0){
    for(T &i:arr){
      array.push_back(vector<T>(0));
      array.end()[-1].push_back(i);
      times.push_back(vector<int>(0));
      times.end()[-1].push_back(t);
    }
  }
 
  void set_vec(vector<T> &arr, int t = 0) {
    for(T &i:arr){
      array.push_back(vector<T>(0));
      array.end()[-1].push_back(i);
      times.push_back(vector<int>(0));
      times.end()[-1].push_back(t);
    }
  }
 
  int size(){
    return (int)array.size();
  }
 
  bool exist(int idx, int t){
    if(idx >= size()) return false;
    return times[idx][0] <= t; 
  }
 
  void add(T &val, int t){
    array.push_back(vector<T>(0));
    array.end()[-1].push_back(val);
    times.push_back(vector<int>(0));
    times.end()[-1].push_back(t);
  }
 
  void update(T &val, int idx, int t){
    array[idx].push_back(val);
    times[idx].push_back(t);
  }
 
  T get(int idx, int t){
    assert(exist(idx,t));
    if(times[idx].end()[-1] <= t){
      return array[idx].end()[-1];
    }
    int i = upper_bound(times[idx].begin(),times[idx].end(),t)-times[idx].begin();
    return array[idx][i-1];
  }
};

File: Data_Structure/PartialPersistentArray.py
# verified with:
# https://atcoder.jp/contests/code-thanks-festival-2017-open/submissions/40153654
# https://atcoder.jp/contests/agc002/submissions/40153678

from typing import TypeVar, List
from bisect import bisect
 
T = TypeVar("T")
class PartialPersistentArray:
    def __init__(self, arr:List[T], t:int = 0) -> None:
        self.array = [[i] for i in arr]
        self.times = [[0] for _ in arr]
 
    def size(self) -> int:
        return len(self.array)
    
    def exist(self, idx:int, t:int) -> bool:
        if idx >= self.size():
            return False
        return self.times[idx][0] <= t
    
    def add(self, val:T, t:int) -> None:
        self.array.append([])
        self.array[-1].append(val)
        self.times.append([])
        self.times[-1].append(t)
 
    def update(self, val:T, idx:int, t:int) -> None:
        self.array[idx].append(val)
        self.times[idx].append(t)
 
    def get(self, idx:int, t:int) -> T:
        if self.times[idx][-1] <= t:
            return self.array[idx][-1]
        return self.array[idx][bisect(self.times[idx],t)-1]